document-legacy/src/document.rs

```diff
@@ -559,50 +621,156 @@ impl Document {
+
+					let mut updated_layer_path = destination_path.clone();
+					updated_layer_path.push(new_layer_id);
+
+					responses.extend([
+						DocumentChanged,
+						CreatedLayer {
+							path: updated_layer_path,
+							is_selected: false,
+						},
+						FolderChanged { path: destination_path.clone() },
+					]);
+				} else if !duplicating {
+					let folder = self.folder_mut(folder_path)?;
+					folder.add_layer(*layer, Some(layer_id), insert_index).ok_or(DocumentError::IndexOutOfBounds)?;
+
+					responses.extend([
+						DocumentChanged,
+						CreatedLayer {
+							path: destination_path.clone(),
+							is_selected: true,
+						},
+						FolderChanged { path: folder_path.to_vec() },
+					]);
+				}
+
+				responses.extend(update_thumbnails_upstream(&destination_path));
 				self.mark_as_dirty(&destination_path)?;
 
-				fn aggregate_insertions(folder: &FolderLayer, path: &mut Vec<LayerId>, responses: &mut Vec<DocumentResponse>) {
+				// Recursively iterate through each layer in a folder and add it to the responses vector
+				fn aggregate_insertions(folder: &FolderLayer, path: &mut Vec<LayerId>, responses: &mut Vec<DocumentResponse>, duplicating: bool) {
 					for (id, layer) in folder.layer_ids.iter().zip(folder.layers()) {
 						path.push(*id);
-						responses.push(DocumentResponse::CreatedLayer { path: path.clone() });
+
+						responses.push(DocumentResponse::CreatedLayer {
+							path: path.clone(),
+							is_selected: !duplicating,
+						});
 						if let LayerDataType::Folder(f) = &layer.data {
-							aggregate_insertions(f, path, responses);
+							aggregate_insertions(f, path, responses, duplicating);
 						}
+
 						path.pop();
 					}
 				}
 
-				let mut responses = Vec::new();
 				if let Ok(folder) = self.folder(&destination_path) {
-					aggregate_insertions(folder, &mut destination_path.as_slice().to_vec(), &mut responses)
+					aggregate_insertions(folder, &mut destination_path.as_slice().to_vec(), &mut responses, duplicating);
 				};
 
-				responses.extend([DocumentChanged, CreatedLayer { path: destination_path.clone() }, FolderChanged { path: folder_path.to_vec() }]);
-				responses.extend(update_thumbnails_upstream(&destination_path));
 				Some(responses)
 			}
 			Operation::DuplicateLayer { path } => {
+				// Notes for review: I wasn't sure on how to apply unwrap_or() to lines of code that use the funciton self.layer()
+
 				let layer = self.layer(&path)?.clone();
 				let (folder_path, _) = split_path(path.as_slice()).unwrap_or((&[], 0));
-				let folder = self.folder_mut(folder_path)?;
 
+				let layer_is_folder = layer.as_folder().is_ok();
+
+				// Recursively collect each of the nested folders and shapes if the layer is a folder
+				let mut duplicated_layers = if layer_is_folder { recursive_collect(self, &path) } else { Vec::new() };
+				// Duplicate the Layer objects for our duplicated layers
```

Comment:
comment on line 695 could be worded a bit more clearly

Reply:
good catch